crates/bevy_ui/src/keyboard_navigation.rs

```diff
@@ -0,0 +1,264 @@
+    if let Some(mut interaction) = focus
+        .entity
+        .and_then(|entity| interactions.get_mut(entity).ok())
+    {
+        interaction.set_if_neq(Interaction::Clicked);
+    }
+}
+
+/// Should the [`end_keyboard_click`] system run?
+pub(crate) fn trigger_click_end(keyboard_input: Res<Input<KeyCode>>) -> bool {
+    keyboard_input.just_released(KeyCode::Space) || keyboard_input.just_released(KeyCode::Return)
+}
+
+/// Reset the clicked state.
+pub(crate) fn end_keyboard_click(mut interactions: Query<&mut Interaction>) {
+    interactions.for_each_mut(|mut interaction| {
+        if *interaction == Interaction::Clicked {
+            // The click was triggered by the keyboard, so it doesn't make sense to go to `Interaction::Hovered`.
+            *interaction = Interaction::None;
+        }
+    });
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::prelude::{ButtonBundle, NodeBundle};
+    use crate::{ui_stack_system, Focusable};
+    use bevy_ecs::prelude::*;
+    use bevy_ecs::system::CommandQueue;
+    use bevy_hierarchy::BuildChildren;
+    use bevy_utils::default;
+
+    #[test]
+    fn keyboard_navigation() {
+        let mut world = World::default();
+
+        let mut schedule = Schedule::new();
+        schedule.add_systems((
+            ui_stack_system,
+            keyboard_navigation_system.after(ui_stack_system),
+            update_focused_state.after(keyboard_navigation_system),
+        ));
+
+        let mut queue = CommandQueue::default();
+        let mut commands = Commands::new(&mut queue, &world);
+        
+        commands.init_resource::<bevy_a11y::Focus>();
+        commands.init_resource::<Input<KeyCode>>();
+        commands.init_resource::<UiStack>();
+
+        let mut children = Vec::new();
+        commands
+            .spawn(NodeBundle::default())
+            .with_children(|parent| {
+                for _child in 0..3 {
+                    children.push(
+                        parent
+                            .spawn(ButtonBundle {
+                                computed_visibility: ComputedVisibility::VISIBLE,
+                                ..default()
+                            })
+                            .id(),
+                    );
+                }
+            });
+        queue.apply(&mut world);
+
+        schedule.run(&mut world);
+        assert!(
+            world.get::<Focusable>(children[0]).unwrap().is_focused(),
+            "navigation should start at the first button"
+        );
+        schedule.run(&mut world);
+        assert!(
+            world.get::<Focusable>(children[1]).unwrap().is_focused(),
+            "navigation should go to the second button"
+        );
+
+        // Simulate pressing shift
+        let mut keyboard_input = world
+            .get_resource_mut::<Input<KeyCode>>()
+            .expect("keyboard input resource");
+        keyboard_input.press(KeyCode::ShiftLeft);
```

Comment:
Where is the `Tab` key input simulated?

Reply:
@TimJentzsch The `keyboard_navigation_system` is always run when the schedule is run. In an actual application this system will only be run when the tab key is pressed.

Reply:
Ah right, makes sense!